### PR TITLE
Add Minimize module with gradient descent and adam implementations

### PR DESCRIPTION
**Pull request by fkm3**

### CODE REVIEW

**Comment by fkm3** [comment]
@silky, if you have already done some work on this, I'm happy to try and merge our ideas.

**Comment by silky** [comment]
@fkm3 amazing; thanks heaps for this, i'll give your version a go! i didn't end up getting a chance to revisit my latest attempt at this (i got completely stuck on the initialised variable thing).

File: tensorflow-ops/src/TensorFlow/Minimize.hs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import qualified TensorFlow.Gradient as TF
+import qualified TensorFlow.Ops as TF hiding (assign, initializedVariable)
+import qualified TensorFlow.Variable as TF
+
+-- | Functions that minimize a loss w.r.t. a set of 'TF.Variable's.
+--
+-- 'Minimizer's are defined as a function of the gradients instead of
+-- the loss so that users can apply transformations to the gradients.
+type Minimizer a =
+    forall m. TF.MonadBuild m =>
```
Comment (judah) [inline tensorflow-ops/src/TensorFlow/Minimize.hs]: Clarify in the comment that the returned op runs a single step of the optimization (rather than the full minimization loop)?

File: tensorflow-mnist/app/Main.hs [hunk 2]
```diff
@@ -15,17 +15,17 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedLists #-}
 
-import Control.Monad (zipWithM, when, forM_)
+import Control.Monad (when, forM_)
```
Comment (blackgnezdo) [inline tensorflow-mnist/app/Main.hs]: Sort since you are touching this line?

File: tensorflow-ops/src/TensorFlow/Minimize.hs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+data AdamConfig = AdamConfig
+    { adamLearningRate :: Float
+    , adamBeta1        :: Float
+    , adamBeta2        :: Float
+    , adamEpsilon      :: Float
+    }
+
+instance Default AdamConfig where
+  def = AdamConfig 0.001 0.9 0.999 1e-8
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Minimize.hs]: Document how the default values were chose in haddock?

File: tensorflow-ops/src/TensorFlow/Variable.hs [hunk 4]
```diff
@@ -133,3 +140,53 @@ assignAdd = assignAdd' id
...
     => OpParams -> Variable a -> Tensor v a -> m ControlNode
 assignAdd' params (Variable h _) v = CoreOps.assignAddVariableOp' params h v
+
+-- | Update '*var' according to the Adam algorithm.
+--
+-- lr_t <- learning_rate * sqrt(1 - beta2^t) / (1 - beta1^t)
+-- m_t <- beta1 * m_{t-1} + (1 - beta1) * g_t
+-- v_t <- beta2 * v_{t-1} + (1 - beta2) * g_t * g_t
+-- variable <- variable - lr_t * m_t / (sqrt(v_t) + epsilon)
+resourceApplyAdam :: forall v'4 v'5 v'6 v'7 v'8 v'9 v'10 t
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Variable.hs]: Are ' here for a particular reason? I'd drop them otherwise.

Comment (judah) [inline tensorflow-ops/src/TensorFlow/Variable.hs]: On second look, you can also probably drop the "forall" clauses from `resourceApplyAdam` and `resourceApplyAdam'` since these functions don't need ScopedTypeVariables.  (Up to you.

File: tensorflow-ops/src/TensorFlow/Minimize.hs (hunk 1 above)
Comment (fkm3) [inline tensorflow-ops/src/TensorFlow/Minimize.hs]: Done.

File: tensorflow-mnist/app/Main.hs (hunk 2 above)
Comment (fkm3) [inline tensorflow-mnist/app/Main.hs]: Done.

File: tensorflow-ops/src/TensorFlow/Minimize.hs (hunk 3 above)
Comment (fkm3) [inline tensorflow-ops/src/TensorFlow/Minimize.hs]: Done.

File: tensorflow-ops/src/TensorFlow/Variable.hs (hunk 4 above)
Comment (fkm3) [inline tensorflow-ops/src/TensorFlow/Variable.hs]: It looks like that because I just copied it from the generated ops. I've cleaned it up to match the usual style.

**Comment by silky** [comment]
oh wow, so much happiness on my part!

thanks for this; the thing that i've been trying to build in tensorflow-haskell finally trains now!!

happy days indeed :)
